File: optimization-greedy-algorithms-problems-plus/tests/job-sequencing.test.ts
```typescript
/**
 * tests/job-sequencing.test.ts
 *
 * Test suite for the Job Sequencing with Deadlines problem.
 */

import { Job, jobSequencing } from '../src/algorithms/job-sequencing';
import { deepClone, shuffleArray } from '../src/utils/helpers';

describe('Job Sequencing with Deadlines', () => {

    // Helper to create a job
    const createJob = (id: string, deadline: number, profit: number): Job => ({ id, deadline, profit });

    // Test case 1: Basic scenario with clear optimal choices
    test('should maximize profit for a standard set of jobs', () => {
        const jobs: Job[] = [
            createJob('a', 2, 100),
            createJob('b', 1, 19),
            createJob('c', 2, 27),
            createJob('d', 1, 25),
            createJob('e', 3, 15),
        ];

        // Expected solution based on manual trace in job-sequencing.ts
        // Sorted by profit desc: a(100,2), c(27,2), d(25,1), b(19,1), e(15,3)
        // Max deadline = 3. Slots: [null, null, null]
        // 1. Process 'a': D=2. Try slot 2. slots[2]=a. Profit=100. Scheduled=['a']
        // 2. Process 'c': D=2. Try slot 2 (taken). Try slot 1. slots[1]=c. Profit=100+27=127. Scheduled=['a','c']
        // 3. Process 'd': D=1. Try slot 1 (taken). Skip.
        // 4. Process 'b': D=1. Try slot 1 (taken). Skip.
        // 5. Process 'e': D=3. Try slot 3. slots[3]=e. Profit=127+15=142. Scheduled=['a','c','e']
        const expectedTotalProfit = 142;
        const expectedScheduledJobs = ['a', 'c', 'e']; // Order can vary if jobs have same profit, but IDs should be present

        const { totalProfit, scheduledJobs } = jobSequencing(jobs);

        expect(totalProfit).toBe(expectedTotalProfit);
        expect(scheduledJobs.sort()).toEqual(expectedScheduledJobs.sort()); // Sort for consistent comparison
        expect(scheduledJobs.length).toBe(expectedScheduledJobs.length);
    });

    // Test case 2: All jobs have the same deadline
    test('should pick highest profit jobs if all have same deadline', () => {
        const jobs: Job[] = [
            createJob('j1', 1, 10),
            createJob('j2', 1, 30),
            createJob('j3', 1, 20),
        ];
        // Sorted by profit: j2(30,1), j3(20,1), j1(10,1)
        // Only one slot available (max deadline is 1). Pick j2.
        const expectedTotalProfit = 30;
        const expectedScheduledJobs = ['j2'];

        const { totalProfit, scheduledJobs } = jobSequencing(jobs);
        expect(totalProfit).toBe(expectedTotalProfit);
        expect(scheduledJobs.sort()).toEqual(expectedScheduledJobs.sort());
        expect(scheduledJobs.length).toBe(expectedScheduledJobs.length);
    });

    // Test case 3: No jobs
    test('should return 0 profit and empty array for no jobs', () => {
        const jobs: Job[] = [];
        const { totalProfit, scheduledJobs } = jobSequencing(jobs);
        expect(totalProfit).toBe(0);
        expect(scheduledJobs).toEqual([]);
    });

    // Test case 4: Single job
    test('should schedule single job if valid', () => {
        const jobs: Job[] = [createJob('single', 1, 50)];
        const { totalProfit, scheduledJobs } = jobSequencing(jobs);
        expect(totalProfit).toBe(50);
        expect(scheduledJobs).toEqual(['single']);
    });

    // Test case 5: Job with deadline 0 (invalid, or effectively 1 if min deadline is 1)
    test('should ignore jobs with deadline 0 or less', () => {
        const jobs: Job[] = [
            createJob('valid', 1, 100),
            createJob('invalid', 0, 50), // Deadline 0 means it cannot be scheduled in slot 1 or later
            createJob('negative', -1, 70),
        ];
        const { totalProfit, scheduledJobs } = jobSequencing(jobs);
        expect(totalProfit).toBe(100);
        expect(scheduledJobs).toEqual(['valid']);
    });

    // Test case 6: More jobs than slots, some jobs skipped
    test('should correctly skip jobs when slots are full', () => {
        const jobs: Job[] = [
            createJob('J1', 3, 100),
            createJob('J2', 1, 10),
            createJob('J3', 2, 20),
            createJob('J4', 1, 5),
            createJob('J5', 3, 30),
            createJob('J6', 2, 15),
        ];
        // Sorted by profit: J1(100,3), J5(30,3), J3(20,2), J6(15,2), J2(10,1), J4(5,1)
        // Max deadline = 3. Slots: [null, null, null]
        // 1. J1(100,3): D=3. Schedule in slot 3. slots=[null,null,J1]. P=100. Sch=['J1']
        // 2. J5(30,3): D=3. Try slot 3 (taken). Try slot 2. Schedule in slot 2. slots=[null,J5,J1]. P=100+30=130. Sch=['J1','J5']
        // 3. J3(20,2): D=2. Try slot 2 (taken). Try slot 1. Schedule in slot 1. slots=[J3,J5,J1]. P=130+20=150. Sch=['J1','J5','J3']
        // All slots (1,2,3) are now taken. Max profit = 150.
        // J6, J2, J4 will be skipped.
        const expectedTotalProfit = 150;
        const expectedScheduledJobs = ['J1', 'J5', 'J3'];

        const { totalProfit, scheduledJobs } = jobSequencing(jobs);
        expect(totalProfit).toBe(expectedTotalProfit);
        expect(scheduledJobs.sort()).toEqual(expectedScheduledJobs.sort());
        expect(scheduledJobs.length).toBe(expectedScheduledJobs.length);
    });

    // Test case 7: Jobs with same profit, different deadlines
    test('should handle jobs with same profit, prioritizing later slots', () => {
        const jobs: Job[] = [
            createJob('A', 2, 50),
            createJob('B', 1, 50),
            createJob('C', 2, 10),
        ];
        // Sorted: A(50,2), B(50,1), C(10,2) (A/B order depends on secondary sort or stability)
        // Let's assume A comes before B if profits are equal for our implementation.
        // Max deadline = 2. Slots: [null, null]
        // 1. A(50,2): D=2. Schedule in slot 2. slots=[null,A]. P=50. Sch=['A']
        // 2. B(50,1): D=1. Schedule in slot 1. slots=[B,A]. P=50+50=100. Sch=['A','B']
        // 3. C(10,2): D=2. Try slot 2 (taken). Try slot 1 (taken). Skip.
        const expectedTotalProfit = 100;
        const expectedScheduledJobs = ['A', 'B'];

        const { totalProfit, scheduledJobs } = jobSequencing(jobs);
        expect(totalProfit).toBe(expectedTotalProfit);
        expect(scheduledJobs.sort()).toEqual(expectedScheduledJobs.sort());
        expect(scheduledJobs.length).toBe(expectedScheduledJobs.length);
    });

    // Test case 8: Large number of random jobs
    test('should work for a large set of random jobs', () => {
        const numJobs = 1000;
        const maxDeadline = 500;
        const maxProfit = 1000;
        const jobs: Job[] = [];
        for (let i = 0; i < numJobs; i++) {
            const deadline = Math.floor(Math.random() * maxDeadline) + 1; // Deadline 1 to maxDeadline
            const profit = Math.floor(Math.random() * maxProfit) + 1; // Profit 1 to maxProfit
            jobs.push(createJob(`J${i}`, deadline, profit));
        }

        const { totalProfit, scheduledJobs } = jobSequencing(jobs);

        expect(totalProfit).toBeGreaterThan(0);
        expect(scheduledJobs.length).toBeLessThanOrEqual(maxDeadline); // Cannot schedule more jobs than slots
        expect(scheduledJobs.length).toBeGreaterThanOrEqual(1); // At least one job should be scheduled
        // Further detailed checks would require complex logic to re-verify optimality,
        // but checking basic properties like profit > 0 and number of jobs <= maxDeadline is good.
    });

    // Test case 9: Input order should not affect the total profit or set of scheduled jobs
    test('should produce the same profit and job set regardless of initial input order', () => {
        const originalJobs: Job[] = [
            createJob('a', 2, 100),
            createJob('b', 1, 19),
            createJob('c', 2, 27),
            createJob('d', 1, 25),
            createJob('e', 3, 15),
        ];
        const expectedTotalProfit = 142;
        const expectedScheduledJobs = ['a', 'c', 'e'];

        const shuffledJobs = shuffleArray(deepClone(originalJobs));

        const { totalProfit, scheduledJobs } = jobSequencing(shuffledJobs);

        expect(totalProfit).toBe(expectedTotalProfit);
        expect(scheduledJobs.sort()).toEqual(expectedScheduledJobs.sort());
        expect(scheduledJobs.length).toBe(expectedScheduledJobs.length);
    });

    // Test case 10: Jobs with high deadlines but low profit
    test('should prioritize high profit jobs even if deadlines are tighter', () => {
        const jobs: Job[] = [
            createJob('J1', 1, 100), // High profit, tight deadline
            createJob('J2', 10, 10), // Low profit, loose deadline
            createJob('J3', 2, 50),
        ];
        // Sorted: J1(100,1), J3(50,2), J2(10,10)
        // Max deadline=10.
        // 1. J1(100,1): Schedule in slot 1. P=100. Sch=['J1']. Slots=[J1, null, ...]
        // 2. J3(50,2): Schedule in slot 2. P=100+50=150. Sch=['J1','J3']. Slots=[J1, J3, ...]
        // 3. J2(10,10): Schedule in slot 10. P=150+10=160. Sch=['J1','J3','J2']. Slots=[J1, J3, ..., J2]
        const expectedTotalProfit = 160;
        const expectedScheduledJobs = ['J1', 'J3', 'J2'];

        const { totalProfit, scheduledJobs } = jobSequencing(jobs);
        expect(totalProfit).toBe(expectedTotalProfit);
        expect(scheduledJobs.sort()).toEqual(expectedScheduledJobs.sort());
        expect(scheduledJobs.length).toBe(expectedScheduledJobs.length);
    });
});
```

File: data-structures-greedy-algorithms-complete/src/test/java/com/greedy/problems/JobSequencingTest.java
```java
package com.greedy.problems;

import com.greedy.problems.JobSequencing.Job;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Test class for JobSequencing.java
 */
public class JobSequencingTest {

    private JobSequencing jobSequencing;

    @BeforeEach
    void setUp() {
        jobSequencing = new JobSequencing();
    }

    @Test
    @DisplayName("Test Case 1: Standard Example")
    void testStandardExample() {
        List<Job> jobs = Arrays.asList(
                new Job("a", 2, 100),
                new Job("b", 1, 10),
                new Job("c", 2, 15),
                new Job("d", 1, 25),
                new Job("e", 3, 30)
        );

        List<Job> selectedJobs = jobSequencing.findMaxProfitJobSequence(jobs);
        int totalProfit = jobSequencing.calculateTotalProfit(selectedJobs);

        // Jobs sorted by profit: a(100), e(30), d(25), c(15), b(10)
        // Max Deadline: 3
        // Slots: [_, _, _, _] (index 0 unused, 1, 2, 3)

        // 1. Process 'a' (D=2, P=100). Try slot 2. Slot 2 is free. Schedule 'a' at slot 2.
        //    Slots: [_, _, a, _]
        // 2. Process 'e' (D=3, P=30). Try slot 3. Slot 3 is free. Schedule 'e' at slot 3.
        //    Slots: [_, _, a, e]
        // 3. Process 'd' (D=1, P=25). Try slot 1. Slot 1 is free. Schedule 'd' at slot 1.
        //    Slots: [_, d, a, e]
        // 4. Process 'c' (D=2, P=15). Try slot 2. Slot 2 taken. Try slot 1. Slot 1 taken. Cannot schedule.
        // 5. Process 'b' (D=1, P=10). Try slot 1. Slot 1 taken. Cannot schedule.

        // Expected scheduled jobs: d (slot 1), a (slot 2), e (slot 3)
        // Total Profit: 25 + 100 + 30 = 155
        // Expected IDs by schedule time: d, a, e

        assertEquals(155, totalProfit);
        assertEquals(3, selectedJobs.size());
        assertEquals(Arrays.asList("d", "a", "e"), selectedJobs.stream().map(Job::getId).collect(Collectors.toList()));
    }

    @Test
    @DisplayName("Test Case 2: Jobs with overlapping deadlines, varying profits")
    void testOverlappingDeadlines() {
        List<Job> jobs = Arrays.asList(
                new Job("J1", 4, 20),
                new Job("J2", 1, 10),
                new Job("J3", 1, 40),
                new Job("J4", 1, 30)
        );

        List<Job> selectedJobs = jobSequencing.findMaxProfitJobSequence(jobs);
        int totalProfit = jobSequencing.calculateTotalProfit(selectedJobs);

        // Sorted by profit: J3(40), J4(30), J1(20), J2(10)
        // Max Deadline: 4
        // Slots: [_,_,_,_,_]

        // 1. Process 'J3' (D=1, P=40). Slot 1 is free. Schedule J3 at slot 1.
        //    Slots: [_, J3, _, _, _]
        // 2. Process 'J4' (D=1, P=30). Slot 1 is taken. Cannot schedule.
        // 3. Process 'J1' (D=4, P=20). Slot 4 is free. Schedule J1 at slot 4.
        //    Slots: [_, J3, _, _, J1]
        // 4. Process 'J2' (D=1, P=10). Slot 1 is taken. Cannot schedule.

        // Expected scheduled jobs: J3 (slot 1), J1 (slot 4)
        // Total Profit: 40 + 20 = 60
        // Expected IDs by schedule time: J3, J1

        assertEquals(60, totalProfit);
        assertEquals(2, selectedJobs.size());
        assertEquals(Arrays.asList("J3", "J1"), selectedJobs.stream().map(Job::getId).collect(Collectors.toList()));
    }

    @Test
    @DisplayName("Test Case 3: All jobs have the same deadline")
    void testSameDeadline() {
        List<Job> jobs = Arrays.asList(
                new Job("A", 1, 10),
                new Job("B", 1, 20),
                new Job("C", 1, 30)
        );

        List<Job> selectedJobs = jobSequencing.findMaxProfitJobSequence(jobs);
        int totalProfit = jobSequencing.calculateTotalProfit(selectedJobs);

        // Sorted by profit: C(30), B(20), A(10)
        // Max Deadline: 1
        // Slots: [_,_]

        // 1. Process 'C' (D=1, P=30). Slot 1 is free. Schedule C at slot 1.
        //    Slots: [_, C]
        // 2. Process 'B' (D=1, P=20). Slot 1 is taken. Cannot schedule.
        // 3. Process 'A' (D=1, P=10). Slot 1 is taken. Cannot schedule.

        // Expected scheduled jobs: C (slot 1)
        // Total Profit: 30

        assertEquals(30, totalProfit);
        assertEquals(1, selectedJobs.size());
        assertEquals(Arrays.asList("C"), selectedJobs.stream().map(Job::getId).collect(Collectors.toList()));
    }

    @Test
    @DisplayName("Test Case 4: No jobs")
    void testEmptyList() {
        List<Job> jobs = Collections.emptyList();
        List<Job> selectedJobs = jobSequencing.findMaxProfitJobSequence(jobs);
        int totalProfit = jobSequencing.calculateTotalProfit(selectedJobs);

        assertTrue(selectedJobs.isEmpty());
        assertEquals(0, totalProfit);
    }

    @Test
    @DisplayName("Test Case 5: Single job")
    void testSingleJob() {
        List<Job> jobs = Collections.singletonList(new Job("X", 5, 50));
        List<Job> selectedJobs = jobSequencing.findMaxProfitJobSequence(jobs);
        int totalProfit = jobSequencing.calculateTotalProfit(selectedJobs);

        assertEquals(50, totalProfit);
        assertEquals(1, selectedJobs.size());
        assertEquals(Arrays.asList("X"), selectedJobs.stream().map(Job::getId).collect(Collectors.toList()));
    }

    @Test
    @DisplayName("Test Case 6: Jobs with same profit, different deadlines")
    void testSameProfitDifferentDeadlines() {
        List<Job> jobs = Arrays.asList(
                new Job("J1", 2, 50),
                new Job("J2", 1, 50)
        );

        List<Job> selectedJobs = jobSequencing.findMaxProfitJobSequence(jobs);
        int totalProfit = jobSequencing.calculateTotalProfit(selectedJobs);

        // Sorted by profit (desc), then deadline (asc): J2(1,50), J1(2,50)
        // Max Deadline: 2
        // Slots: [_,_,_]

        // 1. Process 'J2' (D=1, P=50). Slot 1 is free. Schedule J2 at slot 1.
        //    Slots: [_, J2, _]
        // 2. Process 'J1' (D=2, P=50). Slot 2 is free. Schedule J1 at slot 2.
        //    Slots: [_, J2, J1]

        // Expected scheduled jobs: J2 (slot 1), J1 (slot 2)
        // Total Profit: 50 + 50 = 100

        assertEquals(100, totalProfit);
        assertEquals(2, selectedJobs.size());
        assertEquals(Arrays.asList("J2", "J1"), selectedJobs.stream().map(Job::getId).collect(Collectors.toList()));
    }

    @Test
    @DisplayName("Test Case 7: All jobs can be scheduled")
    void testAllJobsCanBeScheduled() {
        List<Job> jobs = Arrays.asList(
                new Job("A", 1, 10),
                new Job("B", 2, 20),
                new Job("C", 3, 30)
        );

        List<Job> selectedJobs = jobSequencing.findMaxProfitJobSequence(jobs);
        int totalProfit = jobSequencing.calculateTotalProfit(selectedJobs);

        // Sorted by profit: C(30), B(20), A(10)
        // Max Deadline: 3

        // 1. C(3,30) -> Slot 3
        // 2. B(2,20) -> Slot 2
        // 3. A(1,10) -> Slot 1

        // Expected: A, B, C (by schedule time)
        // Total Profit: 10 + 20 + 30 = 60

        assertEquals(60, totalProfit);
        assertEquals(3, selectedJobs.size());
        assertEquals(Arrays.asList("A", "B", "C"), selectedJobs.stream().map(Job::getId).collect(Collectors.toList()));
    }

    @Test
    @DisplayName("Test Case 8: Deadlines larger than number of jobs")
    void testDeadlinesLargerThanN() {
        List<Job> jobs = Arrays.asList(
                new Job("A", 10, 100),
                new Job("B", 20, 50)
        );

        List<Job> selectedJobs = jobSequencing.findMaxProfitJobSequence(jobs);
        int totalProfit = jobSequencing.calculateTotalProfit(selectedJobs);

        // Sorted: A(100), B(50)
        // Max Deadline: 20

        // 1. A(10,100) -> Slot 10
        // 2. B(20,50)  -> Slot 20

        // Expected: A, B (by schedule time)
        // Total Profit: 150
        assertEquals(150, totalProfit);
        assertEquals(2, selectedJobs.size());
        assertEquals(Arrays.asList("A", "B"), selectedJobs.stream().map(Job::getId).collect(Collectors.toList()));
    }
}
```